### DISCUSSION TITLE
Multi-language support: which ones to do next?

### DISCUSSION THREAD

**Discussion by caarlos0** [goreleaser/goreleaser#5367] [Ideas, Issue Triage, and General discussions]
Here's the place for you to share which languages and integrations you think should be added next!

Make sure to check if some suggestion was already given, and give it a thumbs up (and/or comment on the thread).

Cheers!

**Comment by jon4hz** [comment]
Support for python (incl. poetry) would be amazing!

**Reply by ryancurrah** [reply]
Plus one to Python. We love Go Releaser and would like to use it for packing and distributing Python artifacts. Mostly Docker images and PyPi.

**Reply by caarlos0** [reply]
hey!

in the case of python, the build would do nothing, right? 

isn't it basically archiving the source code and publishing to pypi? (+ docker and other features goreleaser allows to use)

**Reply by ryancurrah** [reply]
A Python build typically produces a [wheel](https://peps.python.org/pep-0427/) which is a binary distribution format that’s faster to install and doesn’t require users to compile from source. Python source code is typically compiled into an intermediate bytecode representation, which the Python interpreter then executes in a virtual machine-like environment.

`setuptools` is the long-standing standard tool for building these packages, while modern alternatives like `poetry`, `flit`, and `hatch` handle dependency management and packaging. For example, using `poetry`, running `poetry build` generates a wheel ready for PyPI and other distribution steps.

With GoReleaser’s Docker packaging, once you have a wheel ready, you could `COPY` it into the container image and run `pip install ./your-package-name.whl` to bundle your Python package and its dependencies into a final Docker image.

A wheel package itself contains a project’s dependency information, but not the actual dependency packages themselves. When you run pip install `./your-package-name.whl`, pip uses the wheel’s metadata to locate and install any needed dependencies from PyPI (or another configured index).

**Reply by jon4hz** [reply]
It would be nice if you could also build system packages by using the `nfpm` functions if you want to distribute your app / library natively on linux.

**Reply by caarlos0** [reply]
so, been playing with python's wheel format.

I think that, since its not the same as binary, we might need to add a new artifact type for it (which is fine).

so, my todo list roughly is:

- new artifact type
- builder (supporting pip, uv, poetry, not sure which to default to)
- other pipes such as releases, docker, nfpm, archives, etc will need to add support for wheels, too

main reason is that, nfpm, for example, would have to put the wheel somewhere else I presume? (where?)

it would be awesome if you have more info on how you expect these things to work :)

**Reply by ryancurrah** [reply]
My understanding is that the wheel will be copied into the image or system and pip install is ran pointing to the wheel instead of uploading it to a pypi repository. So in my view it's kind of like a binary.

```console
COPY your_package-0.1-py3-none-any.whl /app/
RUN pip install /app/your_package-0.1-py3-none-any.whl && \
    rm /app/your_package-0.1-py3-none-any.whl
```

**Reply by FollowTheProcess** [reply]
> it would be awesome if you have more info on how you expect these things to work :)

Not OP but I do a fair bit of python as well as Go, in case you weren't aware, the python ecosystem is (slowly) standardising on the following:
- https://peps.python.org/pep-0517/
- https://peps.python.org/pep-0518/
- https://peps.python.org/pep-0621/

Precisely to allow basically any build system to know what to do to build python projects. The recommended way of doing this is to define a PEP-621 compliant `pyproject.toml` (tools like hatch, uv, pdm etc. use this format by default. poetry has only just caught up to supporting this format in v2) that declares a build dependency and build backend like so:

```toml
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"
```

Then `python -m build` (https://github.com/pypa/build) will "just work™" and you'll get a wheel out, that can later be pushed to PyPI etc. Tools often implement their own `build` command e.g. `uv build`, `poetry build` etc. but these for the large part do the same thing as `pypa/build`.

I'm imagining GoReleaser's place in the python ecosystem as more of an orchestrator, tying together the build process with all the extra release goodies that people would otherwise have to implement themselves. Tools like https://github.com/astral-sh/uv handle a lot of the internal "pythony" details (which are v messy/complex and I would advise GoReleaser to stay away from IMO).

My recommendation would be GoReleaser support building:
- Pure python packages (no extension modules - python modules written in Rust/C etc. that must be compiled at build time)
- Defined using a PEP-621 pyproject.toml
- And a PEP-517/518 compliant build backend
- By effectively shelling out to `python -m build`

By doing the above you'll end up producing a standards compliant python wheel, using standard tooling that the rest of the ecosystem knows exactly what to do with. Future enhancements may follow but this is a good place to start IMO

I'm a long time GoReleaser user so I'd be happy to help here :)

**Reply by caarlos0** [reply]
nice! that's good input for sure, thank you for that!


yes, i was thinking about doing the "orchestration" thing, yes.


So, a "regular" python-thing software release would be like:

- python -m build to build the wheel file
- create archives (or just upload the wheel directly to the release?)
- create docker images (copy wheel ...;  run pip install wheel)
- create Github release and what not
- optionally push wheel to PyPi 

is that correct?

**Reply by ryancurrah** [reply]
Yep that seems correct to me.

**Reply by caarlos0** [reply]
another piece of that, how are linux packages generally created in these scenarios? like, where does the wheel go?

**Reply by FollowTheProcess** [reply]
- python -m build to build the wheel file ✅
	- _and_ the sdist, effectively just a tarball of the source code, it's created by default by `python -m build`
- create archives (or just upload the wheel directly to the release?) ✅
	- I tend to just upload the wheel and sdist directly to the release when I do this yeah, `python -m build` will put both into the `./dist` directory so I just upload everything under there normally
- create docker images (copy wheel ...; run pip install wheel) ❓
	- I think this should be optional personally? Python packages don't need images creating, but maybe if you're building an application and want to ship it as an image then yeah sure!
- create Github release and what not ✅
- optionally push wheel to PyPi ✅
	- Typical behaviour is to publish both the wheel and source distribution (sdist), both of which will be created by `python -m build`. The "standard" way to do this is with https://github.com/pypa/twine/. The other thing we might want here is [trusted publishing](https://docs.pypi.org/trusted-publishers/) with OIDC but not sure if that's out of scope or not (maybe a follow up). It's what https://github.com/pypa/gh-action-pypi-publish does

> how are linux packages generally created in these scenarios? like, where does the wheel go?

This I don't know I'm afraid

**Reply by iulian-gm** [reply]
It would be really great to have Python support in GoReleaser 👍. I'm setting up a Go project that will have Python code too, so it would be very useful to use GoReleaser for everything, it's such a fantastic tool. I use uv.

**Reply by caarlos0** [reply]
UV WIP: https://github.com/goreleaser/goreleaser/pull/5652

**Reply by caarlos0** [reply]
Poetry WIP: https://github.com/goreleaser/goreleaser/pull/5734

**Comment by polothy** [comment]
At work, we use Go, Python and TypeScript.

**Reply by caarlos0** [reply]
on typescript, what would you expect goreleaser to do, build-wise?

**Reply by polothy** [reply]
Hrm, I'm probably not the best to answer that - but I feel like `pip` and `npm` sort of just handle everything already so goreleaser would just be a wrapper. I'm not even sure if publishing to GitHub releases and brew tap would be useful for these languages like it is for Go.

**Reply by caarlos0** [reply]
That's my impression too

**Reply by ryancurrah** [reply]
You could upload the resulting wheel to the release.

**Reply by ryancurrah** [reply]
Here is an example release with the wheel https://github.com/google-coral/pycoral/releases/tag/v2.0.0

```console
pip install https://google-coral.github.io/py-repo/tflite-runtime/tflite_runtime-2.5.0.post1-cp39-cp39-linux_x86_64.whl
```

**Comment by caarlos0** [comment]
TypeScript with Bun: ✅

https://goreleaser.com/customization/bun-builds/

**Comment by caarlos0** [comment]
TypeScript with Deno: ✅

https://goreleaser.com/customization/builds/deno/

**Comment by infm5** [comment]
Having support for Dart/Flutter will be great.
It'd solve the pain of distributing apps with homebrew, chocolatey etc and make the process much easier.

Thanks

**Reply by vedantmgoyal9** [reply]
Since, Dart doesn't have cross-compilation support, it is a bit difficult to be added to GoReleaser. However, we can add it to GoReleaser Pro, if there's enough demand for it.

**Reply by cpswan** [reply]
[Dart now has cross-compilation](https://dart.dev/tools/dart-compile#cross-compilation-exe) (at least for a variety of Linux targets).

**Comment by akshaybabloo** [comment]
It will be good to have a `goos` for rust, so that something like this can be done - separate builds for Linux and Windows 

```yaml
# yaml-language-server: $schema=https://goreleaser.com/static/schema.json
version: 2
builds:
  - id: "fs_linux"
    builder: rust
    targets:
      - x86_64-unknown-linux-gnu
    command: build
    goos: [linux]
    flags:
      - "--release"
  - id: "fs_windows"
    builder: rust
    targets:
      - x86_64-pc-windows-msvc
    command: build
    goos: [windows]
    flags:
      - "--release"

archives:
  - id: "fs_windows"
    format_overrides:
      - goos: windows
        formats: ["zip"]

checksum:
  name_template: "checksums.txt"
snapshot:
  version_template: "{{ incpatch .Version }}"
```

**Reply by caarlos0** [reply]
that already works, you don't need to specify goos though

**Comment by caarlos0** [comment]
Python with uv: ✅

https://goreleaser.com/customization/builds/uv/

**Comment by atlet99** [comment]
My bad, I can't read this discussions about multinlang support. So, can you help me with Dart/Flutter ? [Support for publishing Dart/Flutter packages to pub.dev](https://github.com/orgs/goreleaser/discussions/6260)

**Comment by caarlos0** [comment]
[Node.js SEA support is landing 🔜](https://github.com/goreleaser/goreleaser/pull/6579)
